Remove SILGen's PostponedCleanup

aschwaighofer opened the pull request:
Remove usage of PostponedCleanup in SILGen in favor of a mandatory SIL pass that ensures the operand lifetime of ``convert_escape_to_noescape [not_guaranteed] %opd`` instructions for the duration of the convert's uses.

```
  %cvt = convert_escape_to_noescape [not_guaranteed] %opd
  ...
  apply %user(%cvt)
  // <- earliest point of ultimate release on %opd
```

When SILGen emits ``convert_escape_to_noescape [not_guaranteed]
%operand`` instructions it assumes that a later SIL pass -- ClosureLifetimeFixup --
comes along and inserts retain_value/release_value instructions such that
the lifetime of the operand for the duration of the trivial closure
result.

aschwaighofer committed:
SIL: Add an [escaped] attribute to convert_escape_to_noescape instruction

To mark when a user of it is known to escape the value. This happens
with materializeForSet arguments which are captured and used in the
write-back. This means we need to keep the context alive until after
the write-back.

Follow-up patches to fully replace the PostponedCleanup hack in SILGen
by a mandatory SIL transformation pass to guarantee the proper lifetime
will use this flag to be more conservative when extending the lifetime.

The problem:

%pa = partial_apply %f(%some_context)
%cvt = convert_escape_to_noescape [not_guaranteed] [escaped] %pa
%ptr = %materialize_for_set(..., %cvt)
...  write_back
... // <-- %pa needs to be alive until after write_back

aschwaighofer committed:
Add a ClosureLifetimeFixup pass

To replace the code in DI and eventually remove PostponedCleanup in a
follow-up.

When SILGen emits ``convert_escape_to_noescape [not_guaranteed]
%operand`` instructions it assumes that a later SIL pass (this pass)
comes along and inserts retain_value/release_value instructions such that
the lifetime of the operand for the duration of the trivial closure
result.

This commit introduces the pass but does not yet use it.

aschwaighofer committed:
Make some functions static

aschwaighofer committed:
SILGen: Remove PostponedCleanup in favor or the SIL pass that fixes
closure lifetimes.

SILGen will now unconditionally emit

  %cvt = convert_escape_to_noescape [guaranteed] %op

instructions. The mandatory ClosureLifetimeFixup pass ensures that %op's
lifetime spans %cvt's uses.

The code in DefiniteInitialization that handled a subset of cases is
removed.

aschwaighofer committed:
Fix test cases

aschwaighofer committed:
Add tests for ClosureLifetimeFixup

aschwaighofer committed:
Add doc for convert_escape_to_noescape attributes

aschwaighofer commented:
@swift-ci Please test

aschwaighofer commented:
@swift-ci Please test source compatibility

aschwaighofer commented:
@swift-ci Please benchmark

DougGregor commented on docs/SIL.rst:
```diff
@@ -4649,6 +4649,14 @@ Converts an escaping (non-trivial) function type to an ``@noescape`` trivial
+
+A ``convert_escape_to_noescape [escaped]`` indiciates that the result was
+passed to a function (materializeForSet) which escapes the closure in a way not
+expresseed by the convert's users. The mandatory pass must ensure the lifetime
```
Typo “expresseed “

aschwaighofer replied:
Thanks! Fixed.

DougGregor commented on lib/SILGen/SILGenApply.cpp:
```diff
@@ -3695,11 +3697,20 @@ CallEmission::applyNormalCall(SGFContext C) {
   auto mv = callee.getFnValue(SGF, isCurried, borrowedSelf);
 
+  // Materialize for set could temporarily escape its arguments.
+  if (callee.isMaterializeForSet) {
```
It seems unfortunate that we’d have to call out materializeForSet by name here. Is it because of the callback nature of materializeForSet’s writeback, which is breaking some rule that user code would not be permitted to break?

aschwaighofer replied:
Yes, materializeForSet can temporarily escape ``@noescape`` arguments until the writeback.

```
protocol P {
  associatedtype Element
  subscript<U>(a: (Element) -> U, b: (U) -> Element) -> U { get set }
}

func increment<T : P>(p: inout T) where T.Element == String {
  p[{Int($0)!}, {String($0)}] += 1
}
```

I can rename the variable to something else. However, ``materializeForSet`` is the only instance of ``canEscapeArgumentsUntilWriteback`` that I am aware of.

DougGregor commented on lib/SILGen/SILGenDecl.cpp:
```diff
@@ -1289,51 +1289,6 @@ CleanupHandle SILGenFunction::enterDestroyCleanup(SILValue valueOrAddr) {
   return Cleanups.getTopCleanup();
 }
 
-PostponedCleanup::PostponedCleanup(SILGenFunction &sgf, bool recursive)
```
Yay!

aschwaighofer committed:
Fix typo

aschwaighofer commented:
@swift-ci Please smoke test

aschwaighofer committed:
SILVerifier: Verify that convert_escape_to_noescape instructions were lowered by ClosureLifetimeFixup

The [not_guaranteed] and [escaped] attributes should be removed in that
process.

aschwaighofer committed:
Add missing null check

aschwaighofer commented:
@swift-ci Please test

aschwaighofer commented:
Source combat run here: https://ci.swift.org/view/Pull%20Request/job/swift-PR-source-compat-suite/1062/
